### PR TITLE
React native - Use Core

### PR DESCRIPTION
**Pull request by igor-dv**
Issue: #4879 #4011 #4852 #4332 #5249 #5309

RN was a bit behind for a long time with all the new features we did. In order to support all of the good things, we should reuse as much as possible from the core. Now when the manager and preview are separated it's much easier to do.

@Gongreg, we can maybe continue the #4752 here or create a new branch from this one.
Issues we can also finish for the breaking v5 release:
- [x] #4879 
- [x] #4011 ~= #4852
- [x] #4332

## What I did
- moved `src/manager` and `src/preview` to `src/client/*` so it will be similar to other apps
- deleted all the webpack/babel/config related code from rn/server and replaced it with the `dev-server` from core
- ~changed `dev-server` to be able to run the webpack only for manager~
- used `standalone` api to build storybook in a dev mode
- removed `--secured` param (`-s`) and used `--https` like in other apps

- removed build command (it was not doing anything).

TBD:
- [ ] Validate breaking changes
- [x] Check something very useful is not missing now
- [ ] Check what to do now with the `environment` parameter
- [ ] Reflect everything in docs
- [ ] Update migration tutorial

### CODE REVIEW

**Comment by igor-dv** [comment]
Currently, when running an app with expo I am getting the babe6/babel7 mismatch...

**Comment by igor-dv** [comment]
upgrading to sdk31 looks solving the babel issue, but there is something new now =/

**Comment by ndelangen** [comment]
@Gongreg @igor-dv should we perhaps merge this efforts into the ui-overhaul?

This would be a new major release right?

**Comment by igor-dv** [comment]
Tbh I prefer to have a separate branch for v5 that works. And merge UI overhaul there once it is ready. Otherwise we won't be able to relaease alphas to check things

**Comment by Gongreg** [comment]
@igor-dv, solved #4332

**Comment by maxkomarychev** [comment]
hey @igor-dv thanks for this, I'd love to play with your fixes asap even if they aren't merged with the upstream, how I can try this? I tried to using git url in my package json against your branch but it doesn't work, assuming this is because it's monorepo managed with lerna.

**Comment by igor-dv** [comment]
Currently, the branch is broken. We need to fix a few things.

**Comment by rkdavidson** [comment]
Hey @igor-dv , any clarity/direction on the things that need fixed so this can be merged?

**Comment by Gongreg** [comment]
Hey. It should more or less work. We need to update documentation and tests things out mostly

**Comment by Gongreg** [comment]
Hey @igor-dv,
It is needed for react-native-client. But we can do it seperately. :)

**Comment by igor-dv** [comment]
@Gongreg , cool, I think let's do anything we can do separably, it will simplify the merge

**Comment by ndelangen** [comment]
This seems to be an issue:
```
Running smoke test in react_native_scripts
$ storybook start --smoke-test --quiet
info => Loading presets
info => Loading custom addons config.


WARN Unable to close preview build!


WARN FATAL broken build!, will close the process,
WARN Fix the error below and restart storybook.

ModuleBuildError: Module build failed (from /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js):
Error: Plugin/Preset files are not allowed to export objects, only functions. In /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-preset-expo/index.js
    at createDescriptor (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:178:11)
    at items.map (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:109:50)
    at Array.map (<anonymous>)
    at createDescriptors (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:109:29)
    at createPresetDescriptors (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:101:10)
    at presets (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-descriptors.js:47:19)
    at mergeChainOpts (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:320:26)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:283:7
    at buildRootChain (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/config-chain.js:120:22)
    at loadPrivatePartialConfig (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/partial.js:85:55)
    at Object.loadPartialConfig (/Users/dev/Projects/GitHub/storybook/core/node_modules/@babel/core/lib/config/partial.js:110:18)
    at Object.<anonymous> (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:140:26)
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:3:103)
    at _next (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:5:194)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:5:364
    at runLoaders (/Users/dev/Projects/GitHub/storybook/core/node_modules/webpack/lib/NormalModule.js:301:20)
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/Users/dev/Projects/GitHub/storybook/core/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at loader.call.then.err (/Users/dev/Projects/GitHub/storybook/core/node_modules/babel-loader/lib/index.js:51:103)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
##teamcity[testFailed name='react_native_scripts']
##teamcity[testFinished name='react_native_scripts']
##teamcity[testStarted name='react_project' captureStandardOutput='true']
```

**Comment by ndelangen** [comment]
I fixed the issue of the cli test being stuck on running init

Other issue of the smoke-test remains

**Comment by ndelangen** [comment]
@igor-dv It kinda looks like we need to upgrade to babel7 for Rn to make this work?

I'm not sure how it passed before?

Maybe it's the change we did here that is surfacing an existing issue?

**Comment by igor-dv** [comment]
Babel 7 was a part of these changes

**Comment by tmeasday** [comment]
Going to remove this from v5 now the decision has been made to release separately

File: addons/ondevice-knobs/package.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
   "name": "@storybook/addon-ondevice-knobs",
-  "version": "5.0.0-beta.3",
-  "description": "Display storybook story knobs on your device.",
+  "private": true,
+  "version": "5.0.0",
+  "description": "Display storybook story knobs on your deviced.",
```
Comment (shilman) [inline addons/ondevice-knobs/package.json]: deviced typo

**Review by shilman — APPROVED** [review]
LGTM!

File: addons/ondevice-knobs/src/panel.js [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import React from 'react';
 import { View, Text, TouchableOpacity } from 'react-native';
 import PropTypes from 'prop-types';
-
+import { SELECT_STORY, FORCE_RE_RENDER } from '@storybook/core-events';
+import { SET, SET_OPTIONS, RESET, CHANGE, CLICK } from '@storybook/addon-knobs/dist/shared';
```
Comment (shilman) [inline addons/ondevice-knobs/src/panel.js]: We shouldn't be importing from `dist` -- let's export this in `@storybook/addon-knobs/events`?

File: app/react-native-server/readme.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Storybook Server for React Native
+Additional package for @storybook/react-native to support server.
```
Comment (shilman) [inline app/react-native-server/readme.md]: => `to support a web server and browser UI that runs alongside the RN on-device UI`

File: app/react-native/package.json [hunk 4]
```diff
@@ -18,64 +18,28 @@
     "directory": "app/react-native"
   },
   "license": "MIT",
-  "main": "dist/index.js",
+  "main": "src/index.js",
```
Comment (shilman) [inline app/react-native/package.json]: Isn't this the non-transpiled version?

File: app/react-native/readme.md [hunk 5]
```diff
@@ -90,7 +82,7 @@ The following parameters can be passed to the start command:
     host to listen on
 -p, --port <port>
     port to listen on
--s, --secured
+--https
```
Comment (shilman) [inline app/react-native/readme.md]: Needs a note in `MIGRATION.md` and/or should be deprecated with one major version of back-compat

File: app/react-native/src/preview/components/OnDeviceUI/index.js [hunk 6]
```diff
@@ -3,11 +3,11 @@ import PropTypes from 'prop-types';
 import { Keyboard, KeyboardAvoidingView, Platform, Animated, TouchableOpacity } from 'react-native';
 import Events from '@storybook/core-events';
 
-import StoryListView from '../StoryListView';
-import StoryView from '../StoryView';
-import Addons from './addons';
+import StoryListView from '../StoryListView/index';
```
Comment (shilman) [inline app/react-native/src/preview/components/OnDeviceUI/index.js]: Isn't `index` implicit?

File: app/react-native/src/preview/index.js [hunk 7]
```diff
@@ -1,24 +1,25 @@
...
 import addons from '@storybook/addons';
 
 import Events from '@storybook/core-events';
 import Channel from '@storybook/channels';
 import createChannel from '@storybook/channel-websocket';
-import { StoryStore, ClientApi } from '@storybook/core/client';
-import OnDeviceUI from './components/OnDeviceUI';
-import StoryView from './components/StoryView';
+import { StoryStore, ClientApi } from '@storybook/client-api';
+import OnDeviceUI from './components/OnDeviceUI/index';
```
Comment (shilman) [inline app/react-native/src/preview/index.js]: index implicit

File: app/react-native/src/preview/index.js [hunk 8]
```diff
@@ -132,10 +134,19 @@ export default class Preview {
     };
   }
 
+  _getHost(params = {}) {
+    if (params.host)
+      return params.host;
+    if (Platform.OS === 'android')
+      return '10.0.2.2';
```
Comment (shilman) [inline app/react-native/src/preview/index.js]: Is this documented somewhere?

File: examples-native/crna-kitchen-sink/package.json [hunk 9]
```diff
@@ -34,18 +36,22 @@
     "@storybook/client-logger": "file:../../packs/storybook-client-logger.tgz",
     "@storybook/components": "file:../../packs/storybook-components.tgz",
     "@storybook/router": "file:../../packs/storybook-router.tgz",
-    "@storybook/core": "file:../../packs/storybook-core.tgz",
+    "@storybook/core": "file:../../lib/core",
     "@storybook/core-events": "file:../../packs/storybook-core-events.tgz",
     "@storybook/node-logger": "file:../../packs/storybook-node-logger.tgz",
     "@storybook/react-native": "file:../../packs/storybook-react-native.tgz",
+    "@storybook/react-native-server": "file:../../app/react-native-server",
```
Comment (shilman) [inline examples-native/crna-kitchen-sink/package.json]: Why not `.tgz`? Can we remove the other `.tgz` files and remove the `build-packs` step entirely?

File: lib/cli/bin/generate.js [hunk 10]
```diff
@@ -22,6 +22,7 @@ if (process.argv[1].includes('getstorybook')) {
     .option('-N --use-npm', 'Use npm to install deps')
     .option('-p --parser <babel | babylon | flow>', 'jscodeshift parser')
     .option('-t --type <type>', 'Add Storybook for a specific project type')
+    .option('-y --yes', 'Answer yes to all prompts')
```
Comment (shilman) [inline lib/cli/bin/generate.js]: Should this be a separate PR?

**Review by shilman — CHANGES_REQUESTED** [review]
So happy this is almost mergeable. Some minor nitpicks above, but otherwise it looks great to me.

File: app/react-native/src/preview/index.js (hunk 7 above)
Comment (ndelangen) [reply app/react-native/src/preview/index.js]: Right, but makes it harder to find, if neglected.

File: app/react-native/src/preview/components/OnDeviceUI/index.js (hunk 6 above)
Comment (ndelangen) [reply app/react-native/src/preview/components/OnDeviceUI/index.js]: Yes, but I don't recommend relying on that behavior

File: app/react-native/readme.md (hunk 5 above)
Comment (ndelangen) [reply app/react-native/readme.md]: I will add it to the MIGRATION.MD

File: app/react-native/src/preview/components/OnDeviceUI/index.js (hunk 6 above)
Comment (shilman) [reply app/react-native/src/preview/components/OnDeviceUI/index.js]: Why not? Doesn't the majority of JS code (including most of the storybook codebase) rely on that?

Comment (benoitdion) [reply app/react-native/src/preview/components/OnDeviceUI/index.js]: removed the `/index` since it is implicit and is consistent with the rest of the codebase

**Review by shilman — APPROVED** [review]
Awesome!!!

**Review by ndelangen — APPROVED** [review]

**Comment by igor-dv** [comment]
Oops I am late. Yay !! Great Success 💥

**Comment by dangreenisrael** [comment]
🎉
